### PR TITLE
Add dictionaries cloud sync tests

### PR DESCRIPTION
**Pull request by clementetb**
Adds dictionary fields into sync objects to validate correct schema transfer.

It also validates a roundtrip of data with all the dictionary's supported types.

⚠️ Not working. Sync cloud test server not ready.

### CODE REVIEW

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt [hunk 1]
```diff
@@ -554,8 +578,10 @@ class SyncedRealmTests {
                     columnDecimal128 = expectedDecimal128
                     columnObjectId = expectedObjectId
                     columnUUID = expectedUUID
+                    columnMixed = expectedMixed
```
Comment (rorbech) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt:581]: Just for completion I think we should just iterate all the types and see if they sync properly for a single field. We could just add a loop with a `realm1sesseion.upload...()` followed by `realm2session.download...() at the end of this test.

**Review by rorbech — CHANGES_REQUESTED** [review]
Just proposed a small addition to the test for completion. ...when the server is fixed.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt [hunk 2]
```diff
@@ -655,6 +680,18 @@ class SyncedRealmTests {
                         assertEquals(expectedObjectIdList, it.columnObjectIdList)
                         assertEquals(expectedUUIDList, it.columnUUIDList)
                         assertEquals(expectedMixedList, it.columnMixedList)
+
+                        assertEquals(expectedObjectId, it.columnRealmDictionary["key"]!!.id)
```
Comment (rorbech) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt:684]: I know that it seems a bit boiler plate-ish to verify that the tests are exhaustive, but it is difficult to verify that you have covered all types from the above. Ex. `RealmDictionary<SyncDog>` seems to be left out.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt [hunk 3]
```diff
@@ -655,6 +680,18 @@ class SyncedRealmTests {
...
+                        assertEquals(expectedString, it.columnStringDictionary["key"])
+                        assertTrue(Arrays.equals(expectedBinary, it.columnBinaryDictionary["key"]))
+                        assertEquals(expectedBoolean, it.columnBooleanDictionary["key"])
+                        assertEquals(expectedLong, it.columnLongDictionary["key"])
+                        assertEquals(expectedDouble, it.columnDoubleDictionary["key"])
+                        assertEquals(expectedDate, it.columnDateDictionary["key"])
+                        assertEquals(expectedDecimal128, it.columnDecimal128Dictionary["key"])
+                        assertEquals(expectedObjectId, it.columnObjectIdDictionary["key"])
+                        assertEquals(expectedUUID, it.columnUUIDDictionary["key"])
+                        assertEquals(expectedMixed, it.columnMixedDictionary["key"])
```
Comment (rorbech) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt:694]: Does this pass? Or did you just not get to it yet due to the missing cloud support. I think I saw a notice of the issues of dictionary with nullable links being fixed in the latest server image, so maybe worth a respin.

**Review by rorbech — COMMENTED** [review]
Minor questions on test coverage.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt (hunk 2 above)
Comment (clementetb) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt]: Have updated with an exhaustive switch-on-type.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt (hunk 3 above)
Comment (clementetb) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt]: Today's image doesn't contain the fix.

**Pull request merged**